src/lib.rs

```diff
@@ -449,28 +441,37 @@ where
 	let mut validation_result =
 		CommitValidationResult { num_precommits: commit.precommits.len(), ..Default::default() };
 
-	// check that all precommits are for blocks higher than the target
-	// commit block, and that they're its descendents
-	let all_precommits_higher_than_target = commit.precommits.iter().all(|signed| {
-		signed.precommit.target_number >= commit.target_number &&
-			chain.is_equal_or_descendent_of(
-				commit.target_hash.clone(),
-				signed.precommit.target_hash.clone(),
-			)
+	// the base of the round should be the lowest block for which we can find a
+	// precommit (any vote would only have been accepted if it was targetting a
+	// block higher or equal to the round base)
+	let base = match commit
+		.precommits
+		.iter()
+		.map(|signed| &signed.precommit)
+		.min_by_key(|precommit| precommit.target_number)
+		.map(|precommit| (precommit.target_hash.clone(), precommit.target_number))
+	{
+		None => return Ok(validation_result),
+		Some(base) => base,
+	};
+
+	// check that all precommits are for blocks that are equal to or descendants
+	// of the round base
+	let all_precommits_higher_than_base = commit.precommits.iter().all(|signed| {
```

Comment:
Is the number of invalid voters ( in order to initialize `num_invalid_voters`) deliberately ignored?
Currently is left to 0 (default).
I don't think is paramount, and indeed looks like the value is just ignored by Substrate

Reply:
Currently the semantics for `num_invalid_voters` is votes from voters that are not part of the validator set, i.e. a message signed by some random key, so it should be OK to stay at 0 here (we didn't check so we don't know).